[Bug]: Error occurs when using Eagle3: Encoder cache miss for {mm_hash}

### Your current environment

<details>
<summary>The output of <code>python collect_env.py</code></summary>


```text
Your output of `python collect_env.py` here
```
==============================
        System Info
==============================
OS                           : CentOS Linux 7 (Core) (x86_64)
GCC version                  : (GCC) 12.1.0
Clang version                : Could not collect
CMake version                : version 3.31.1
Libc version                 : glibc-2.33

==============================
       PyTorch Info
==============================
PyTorch version              : 2.9.0+cu128
Is debug build               : False
CUDA used to build PyTorch   : 12.8
ROCM used to build PyTorch   : N/A

==============================
      Python Environment
==============================
Python version               : 3.10.12 (main, Jan 16 2025, 13:28:48) [GCC 12.1.0] (64-bit runtime)
Python platform              : Linux-5.10.0-1.0.0.39-x86_64-with-glibc2.33

==============================
       CUDA / GPU Info
==============================
Is CUDA available            : True
CUDA runtime version         : 12.8.61
CUDA_MODULE_LOADING set to   :
GPU models and configuration :
GPU 0: NVIDIA L20
GPU 1: NVIDIA L20
GPU 2: NVIDIA L20
GPU 3: NVIDIA L20

Nvidia driver version        : 535.161.08
cuDNN version                : Could not collect
HIP runtime version          : N/A
MIOpen runtime version       : N/A
Is XNNPACK available         : True

==============================
          CPU Info
==============================
Architecture:          x86_64
CPU op-mode(s):        32-bit, 64-bit
Byte Order:            Little Endian
CPU(s):                128
On-line CPU(s) list:   0-127
Thread(s) per core:    2
Core(s) per socket:    32
Socket(s):             2
NUMA node(s):          2
Vendor ID:             GenuineIntel
CPU family:            6
Model:                 106
Model name:            Intel(R) Xeon(R) Platinum 8350C CPU @ 2.60GHz
Stepping:              6
CPU MHz:               3100.000
CPU max MHz:           3500.0000
CPU min MHz:           800.0000
BogoMIPS:              5200.00
Virtualization:        VT-x
L1d cache:             48K
L1i cache:             32K
L2 cache:              1280K
L3 cache:              49152K
NUMA node0 CPU(s):     0-31,64-95
NUMA node1 CPU(s):     32-63,96-127
Flags:                 fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect wbnoinvd dtherm ida arat pln pts avx512vbmi umip pku avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq rdpid fsrm md_clear pconfig flush_l1d arch_capabilities

==============================
Versions of relevant libraries
==============================
[pip3] flashinfer-python==0.5.2
[pip3] numpy==2.2.6
[pip3] nvidia-cublas-cu12==12.8.4.1
[pip3] nvidia-cuda-cupti-cu12==12.8.90
[pip3] nvidia-cuda-nvrtc-cu12==12.8.93
[pip3] nvidia-cuda-runtime-cu12==12.8.90
[pip3] nvidia-cudnn-cu12==9.10.2.21
[pip3] nvidia-cudnn-frontend==1.16.0
[pip3] nvidia-cufft-cu12==11.3.3.83
[pip3] nvidia-cufile-cu12==1.13.1.3
[pip3] nvidia-curand-cu12==10.3.9.90
[pip3] nvidia-cusolver-cu12==11.7.3.90
[pip3] nvidia-cusparse-cu12==12.5.8.93
[pip3] nvidia-cusparselt-cu12==0.7.1
[pip3] nvidia-cutlass-dsl==4.3.0
[pip3] nvidia-ml-py==13.580.82
[pip3] nvidia-nccl-cu12==2.27.5
[pip3] nvidia-nvjitlink-cu12==12.8.93
[pip3] nvidia-nvshmem-cu12==3.3.20
[pip3] nvidia-nvtx-cu12==12.8.90
[pip3] pyzmq==27.0.2
[pip3] torch==2.9.0
[pip3] torchaudio==2.9.0
[pip3] torchvision==0.24.0
[pip3] transformers==4.57.1
[pip3] triton==3.5.0
[conda] Could not collect

==============================
         vLLM Info
==============================
ROCM Version                 : Could not collect
vLLM Version                 : 0.11.3.dev0+g275de3417.d20260106 (git sha: 275de3417, date: 20260106)
vLLM Build Flags:
  CUDA Archs: Not Set; ROCm: Disabled
GPU Topology:
  	GPU0	GPU1	GPU2	GPU3	NIC0	CPU Affinity	NUMA Affinity	GPU NUMA ID
GPU0	 X 	NODE	SYS	SYS	NODE	0-31,64-95	0		N/A
GPU1	NODE	 X 	SYS	SYS	NODE	0-31,64-95	0		N/A
GPU2	SYS	SYS	 X 	NODE	SYS	32-63,96-127	1		N/A
GPU3	SYS	SYS	NODE	 X 	SYS	32-63,96-127	1		N/A
NIC0	NODE	NODE	SYS	SYS	 X

Legend:

  X    = Self
  SYS  = Connection traversing PCIe as well as the SMP interconnect between NUMA nodes (e.g., QPI/UPI)
  NODE = Connection traversing PCIe as well as the interconnect between PCIe Host Bridges within a NUMA node
  PHB  = Connection traversing PCIe as well as a PCIe Host Bridge (typically the CPU)
  PXB  = Connection traversing multiple PCIe bridges (without traversing the PCIe Host Bridge)
  PIX  = Connection traversing at most a single PCIe bridge
  NV#  = Connection traversing a bonded set of # NVLinks

NIC Legend:

  NIC0: mlx5_bond_0

==============================
     Environment Variables
==============================
NCCL_SOCKET_IFNAME=bond0
CUDA_HOME=/home/disk3/hzh/cuda-12.8
CUDA_HOME=/home/disk3/hzh/cuda-12.8
LD_LIBRARY_PATH=/home/disk3/hzh/cuda_driver:/home/disk3/hzh/python-env/zlib/lib:/home/disk3/hzh/python-env/libffi/lib:/home/disk3/hzh/python-env/openssl/lib:/home/disk3/hzh/python-env/sqlite/lib:/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/nvidia/nvjitlink/lib/:/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib:/home/disk3/work/openmpi-4.1.6//lib/:/home/disk3/hzh/nccl_2.26.2-1+cuda12.8_x86_64/lib/:/home/disk3/hzh/cudnn-9.11.1-for-cuda-12/lib/:/home/disk3/hzh/cuda-12.8//targets/x86_64-linux/lib:/home/disk3/hzh/cuda-12.8//lib64:/home/disk3/hzh/cuda-12.8//lib:/usr/lib64/libcuda.so.1:/home/disk3/hzh/cuda_driver:/home/disk3/hzh/python-env/zlib/lib:/home/disk3/hzh/python-env/libffi/lib:/home/disk3/hzh/python-env/openssl/lib:/home/disk3/hzh/python-env/sqlite/lib:/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/nvidia/nvjitlink/lib/:/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib:/home/disk3/work/openmpi-4.1.6//lib/:/home/disk3/hzh/nccl_2.26.2-1+cuda12.8_x86_64/lib/:/home/disk3/hzh/cudnn-9.11.1-for-cuda-12/lib/:/home/disk3/hzh/cuda-12.8//targets/x86_64-linux/lib:/home/disk3/hzh/cuda-12.8//lib64:/home/disk3/hzh/cuda-12.8//lib:/usr/lib64/libcuda.so.1:
CUDA_VISIBLE_DEVICES=2,3
CUDA_VISIBLE_DEVICES=2,3
TORCH_NVCC_FLAGS=-D_GLIBCXX_USE_CXX11_ABI=1
NCCL_P2P_DISABLE=0
NCCL_HOME=/home/disk3/hzh/nccl_2.26.2-1+cuda12.8_x86_64
NCCL_IB_DISABLE=1
CUDA_AND_CUDNN_LD_LIBRARY_PATH=/home/disk3/hzh/cuda-12.8:/home/disk3/hzh/cudnn-9.11.1-for-cuda-12
CUDNN_HOME=/home/disk3/hzh/cudnn-9.11.1-for-cuda-12
PYTORCH_NVML_BASED_CUDA_CHECK=1
TORCHINDUCTOR_COMPILE_THREADS=1



</details>


### 🐛 Describe the bug

When using speculative decoding (EAGLE3) together with multimodal (MM) inputs, the model may attempt to access MM encoder cache entries using speculative positions, and hit Encoder cache miss assertions. My model is Qwen2.5-VL-72b and the draft model is trained via SpecForge.
This happens even when all requests are processed serially, with no concurrency.
And I noticed that thie occurs only when start_pos = num_computed_tokens + num_scheduled_tokens - 1.

The detailed tracback shown below:
`(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815] WorkerProc hit an exception.
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815] Traceback (most recent call last):
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/executor/multiproc_executor.py", line 810, in worker_busy_loop
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     output = func(*args, **kwargs)
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return func(*args, **kwargs)
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_worker.py", line 539, in sample_tokens
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return self.model_runner.sample_tokens(grammar_output)
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return func(*args, **kwargs)
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2982, in sample_tokens
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     propose_draft_token_ids(sampled_token_ids)
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2946, in propose_draft_token_ids
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     self._draft_token_ids = self.propose_draft_token_ids(
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 3240, in propose_draft_token_ids
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     mm_embed_inputs = self._gather_mm_embeddings(
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815] WorkerProc hit an exception.
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2058, in _gather_mm_embeddings
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815] Traceback (most recent call last):
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     assert encoder_output is not None, f"Encoder cache miss for {mm_hash}."
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/executor/multiproc_executor.py", line 810, in worker_busy_loop
(Worker_TP1 pid=120245) ERROR 01-16 17:46:35 [multiproc_executor.py:815] AssertionError: Encoder cache miss for b9e74b753b4e47844c03af5e0092ef9f413e5e7f5aaccaa5b1427a8457296add.
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     output = func(*args, **kwargs)
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return func(*args, **kwargs)
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_worker.py", line 539, in sample_tokens
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return self.model_runner.sample_tokens(grammar_output)
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/torch/utils/_contextlib.py", line 120, in decorate_context
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     return func(*args, **kwargs)
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2982, in sample_tokens
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     propose_draft_token_ids(sampled_token_ids)
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2946, in propose_draft_token_ids
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     self._draft_token_ids = self.propose_draft_token_ids(
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 3240, in propose_draft_token_ids
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     mm_embed_inputs = self._gather_mm_embeddings(
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]   File "/home/disk3/hzh/python-env/python-3.10.12-vllm0.11.2/lib/python3.10/site-packages/vllm/v1/worker/gpu_model_runner.py", line 2058, in _gather_mm_embeddings
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815]     assert encoder_output is not None, f"Encoder cache miss for {mm_hash}."
(Worker_TP0 pid=120244) ERROR 01-16 17:46:35 [multiproc_executor.py:815] AssertionError: Encoder cache miss for b9e74b753b4e47844c03af5e0092ef9f413e5e7f5aaccaa5b1427a8457296add.`

### Before submitting a new issue...

- [x] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

--- a/tests/v1/core/test_scheduler.py
+++ b/tests/v1/core/test_scheduler.py
@@ -3675,3 +3675,72 @@ def test_abort_request_finished_recving():
     # verify request is deleted
     assert request.request_id not in scheduler.requests
     assert not scheduler.finished_recving_kv_req_ids
+
+
+def test_eagle3_mm_encoder_cache_with_shift():
+    """Test EAGLE3 encoder scheduling accounts for shift_computed_tokens.
+
+    Regression test for issue #32469: When EAGLE3 is enabled with
+    disable_chunked_mm_input=True, ensure encoder inputs are scheduled
+    when tokens overlap the MM range, properly accounting for
+    shift_computed_tokens in the boundary calculation.
+
+    Without the fix, the scheduler would fail to schedule encoder inputs
+    at the boundary, causing "Encoder cache miss" errors.
+    """
+    scheduler = create_scheduler(
+        model="llava-hf/llava-1.5-7b-hf",
+        max_num_batched_tokens=1024,
+        disable_chunked_mm_input=True,
+        max_model_len=2048,
+        num_speculative_tokens=4,  # This enables EAGLE with shift=1
+    )
+
+    mm_start_pos = 100
+    mm_length = 576
+
+    mm_positions = [
+        [PlaceholderRange(offset=mm_start_pos, length=mm_length)],
+    ]
+
+    requests = create_requests(
+        num_requests=1,
+        num_tokens=mm_start_pos + mm_length + 100,
+        mm_positions=mm_positions,
+    )
+
+    # Start with some tokens already computed to simulate decoding
+    request = requests[0]
+    request.num_computed_tokens = 0
+
+    scheduler.add_request(request)
+    output = scheduler.schedule()
+
+    assert output is not None
+    shift_computed_tokens = 1
+    req_id = request.request_id
+
+    assert req_id in output.num_scheduled_tokens
+    num_scheduled = output.num_scheduled_tokens[req_id]
+
+    mm_feature = request.mm_features[0]
+    start_pos = mm_feature.mm_position.offset
+    tokens_end = request.num_computed_tokens + num_scheduled
+    scheduled_end_with_shift = tokens_end + shift_computed_tokens
+
+    # Assert that we scheduled into the MM range (test setup verification)
+    assert scheduled_end_with_shift > start_pos, (
+        f"Test setup error: expected to schedule into MM range. "
+        f"scheduled_end_with_shift={scheduled_end_with_shift}, "
+        f"start_pos={start_pos}"
+    )
+
+    # The key assertion: when scheduled tokens overlap MM range
+    # (accounting for EAGLE's shift), encoder MUST be scheduled.
+    # Without the fix, this would fail at the boundary case.
+    assert req_id in output.scheduled_encoder_inputs, (
+        f"Encoder input missing: scheduled {num_scheduled} tokens "
+        f"(computed={request.num_computed_tokens}, end={tokens_end}, "
+        f"shifted_end={scheduled_end_with_shift}) overlapping MM at "
+        f"{start_pos}. The fix must schedule encoder inputs."
+    )

--- a/vllm/v1/core/sched/scheduler.py
+++ b/vllm/v1/core/sched/scheduler.py
@@ -1155,7 +1155,12 @@ class Scheduler(SchedulerInterface):
                 and (num_computed_tokens + num_new_tokens)
                 < (start_pos + num_encoder_tokens)
             ):
-                num_new_tokens = start_pos - num_computed_tokens
+                # Account for EAGLE shift when rolling back to avoid
+                # encoder cache miss. This ensures the scheduled range
+                # stops before start_pos even with the shift.
+                num_new_tokens = max(
+                    0, start_pos - (num_computed_tokens + shift_computed_tokens)
+                )
                 break
             if not self.encoder_cache_manager.can_allocate(
                 request, i, encoder_compute_budget, num_embeds_to_schedule

